### ISSUE TITLE
Comments in lines

### ISSUE THREAD

**Issue by monsieurhannes** [nextcloud/text#185]
**Is your feature request related to a problem? Please describe.**
No, but it would be great to have the ability to comment text directly in the line.

**Describe the solution you'd like**
Similar to gdocs and word, marking a text and writing a comment on the side.

**Describe alternatives you've considered**
Linking a specific part of the text in the comment section in the panel on the right could also be a solution, maybe for the start, being able to link headlines `#H2`

Just an idea, might be too ambitious, but maybe sth. for the backlogs.

**Comment by jancborchardt** [comment]
@monsieurhannes yes, this would be very useful. Could you show screenshots from how Google Docs, Word and others do it as examples?

**Comment by juliushaertl** [comment]
Besides the frontend we of course need to think of a proper way how we represent the comments in markdown.

**Comment by tiotrom** [comment]
This would be such an awesome thing

**Comment by juliushaertl** [comment]
Please use the [Github Reactions](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) feature instead of commenting that also need a specific feature. That way we can see your support without notifying all users subscribed to a given issue.

**Comment by joshp23** [comment]
Markdown [footnotes](https://www.markdownguide.org/extended-syntax#footnotes) would be a possibility to represent in MD

**Comment by jtoloe** [comment]
HackMD has this functionality. See https://hackmd.io/s/how-to-use-comments for examples. I do not think that the comments are stored in the markdown file.

**Comment by juliushaertl** [comment]
The issue with not storing it in the markdown file would be that we then loose the comments as soon as someone changes the file from outside (e.g. with a editor synced by the desktop client)

**Comment by juliushaertl** [comment]
Footnotes would be a possible way, but they limit comments to one cursor position in the text rather than having the possibility to mark a range of text and comment on that.

**Comment by jnnkB** [comment]
Maybe Critic Markup would be an option: http://criticmarkup.com/users-guide.php

**Comment by monsieurhannes** [comment]
Any development on this front? 
It seems like it gets mentioned more often recently. I guess Text is in a bit of competition with the big office suites which provide that feature already and maybe it's an overkill for Text. But at the same time Text is much more lightweight and accessible on all platforms.
I'd be happy to help in any way - more in the UI than the programming part. 😆

**Comment by juliushaertl** [comment]
This would still be a nice to have feature, but is currently blocked by a proper approach to persist comments in markdown files, without impacting readability too much. 

The only feasible option seems to be to use footnotes for this, but then you could only comment on cursor positions and not comment on a selected area.

**Comment by susnux** [comment]
I think range based comments are only possible with custom syntax.
I would not use footnotes for this, as they might be added as a stand-alone feature in the future.

When talking about inline comments, how about HTML comment blocks?
They are allowed as per markdown spec.

```markdown
# Header
some text <!-- some comment --> text
```

But when talking about range comments, then I could not even find one already used syntax for that.
So to stick with current style of markdown syntax this could be one option:

```markdown
# Header
?[some text](with comment)
```

**Comment by badrihippo** [comment]
## Comments as per CriticMarkup

Since many people have been mentioning CriticMarkup in this and duplicate threads, here's how it looks in practice.

Comment (sticks to a specific point in the text `{>> like this <<}`):
```
Lorem ipsum dolor sit amet.{>>This is a comment<<}
```

Highlight (highlights `{==a specific passage==}{>>like this<<}`):
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit. {==Vestibulum at
orci magna. Phasellus augue justo, sodales eu pulvinar ac, vulputate eget
nulla.==}{>>confusing<<} Mauris massa sem, tempor sed cursus et, semper
tincidunt lacus.
```

To me this looks like an elegant solution, which will preserve the comments even when edited with an external editor.

One caveat is that the CriticMarkup parser doesn't currently support broken markdown tags: a highlight has to include both the opening and closing of the tag to work properly. They hope to fix this in the future and allow partial tags to be highlighted too.

## First steps

I think even the first option (just inline comments at a specific point in the text, not necessarily tied to a specific highlighted passed) will be a great help for collaborative editing. Given that it also looks easier to implement, how about we start with that, and put highlights as a feature later down the line if needed?

## Linking to the user

What remains to be standardised is how to link the comment to a NextCloud account. The simplest way is to prepend the commenter's name before adding the actual comment, like this:

```
NextCloud has a great collaborative editor {>>>badrihippo@domain.tld: and now we have comments!<<<}
```

If we're only saving these comments directly to the Markdown file, there is obviously no way to verify the identity: anyone can type my username into the text file to make it appear as if I wrote the comment. To prevent this, perhaps we could implement this in multiple stages:

1. Start with identity-less comments: have them simply show up as speech bubbles without specifying who wrote the comment.
2. Start specifying the comment author with the above syntax, but also save the comment information externally so that it can be cross-checked. If the comment doesn't match, we could either remove it (destructive) or display a warning saying that the identity of the commenter could not be verified. (I prefer the second option)
3. (Not sure how feasible this is) scan the Markdown file everytime it is edited, so that if the user has manually inserted a comment with the given syntax, and under their username, then it gets added to the database as "their" comment. This goes hand-in-hand with the "could not be verified" comment handling behaviour

---

That's as far as I've thought so far—is there anything I missed, any concerns, or anything that could be done better? (Tagging @juliushaertl for what you think of the markdown mechanics, and if you have any concerns about using CriticMarkup instead of vanilla markdown or whatever we use right now).

**Comment by mejo-** [comment]
Thanks a lot @badrihippo for the thorough reply. I like your approach. [CriticMarkup](https://github.com/CriticMarkup/CriticMarkup-toolkit) doesn't look very active, but the syntax seems sensible nevertheless.

And there is a [markdown-it plugin for CriticMarkup](https://github.com/wafer-li/markdown-it-criticmarkup) (that is very old but then it's only a few lines of code anyway). We would need that for serializing the markdown into HTML.

There's also [vimtaai/critic-markup](https://github.com/vimtaai/critic-markup/) which might help as code inspiration (e.g. for parsing regexes).

**Comment by pierreozoux** [comment]
This discussion looks a lot like this one:
https://github.com/hedgedoc/hedgedoc/issues/2879

I would love that `nextcloud/text` and `hedgedoc` agree on the same solution so that I could use hedgedoc and other people text, and this would be compatible :)

It is probably utopia, but this is the way I feel, a common internal standard to collaborate between different tools, and different UI according to preferences.

Sorry to open the scope of the discussion, don't hesitate to label it as offtopic.

Have a nice day!

**Comment by codedust** [comment]
To get a better feeling for this solution, here are some screenshots of popular text editors, as @jancborchardt suggested:

### Google Docs
![grafik](https://github.com/nextcloud/text/assets/7175914/a472b30e-101c-466f-84de-13e2cbb64bda)

### LibreOffice Text
![grafik](https://github.com/nextcloud/text/assets/7175914/7f5e7e73-712f-4dae-b020-0667c454ecc5)

### HackMD
![grafik](https://github.com/nextcloud/text/assets/7175914/86a00ae3-d14d-4fff-a18c-b6d65ee2806b)

see https://hackmd.io/s/how-to-use-comments

### OnlyOffice
![grafik](https://github.com/nextcloud/text/assets/7175914/7a41b847-528a-4097-a80f-69d2dd02bbf6)

**Comment by micimize** [comment]
I like the idea in https://github.com/hedgedoc/hedgedoc/issues/2879#issuecomment-1324237737 of using in-doc nested blockquotes as the fundamental data structure, but I'd suggest:
1.  Adding an annotation like `[style=floating]` and `[references=^1]` so that something like `==comment target[^1]==` could be an unambiguous target and the floating sidebar style could be selected by an editor
2. Adding an easy "highlight and comment" affordance that handles the annotations
3. Auto-collapse comment threads in the editor view.

Not entirely sure of the implementation details but `[style=floating]` could be a generic "block" annotation & the rendered sidebar would just be markdown with some special css. ideally I think the bubbles could even optionally annotate raw editor view.

**Comment by max-nextcloud** [comment]
I played with some plain commonmark markdown syntax. Actually it's quite difficult to come up with something that will work if multiple comments exist for overlapping ranges.

Using links with titles would be nice in terms of the rendering outside text - but it does not allow editing in other editors that do not allow editing the underlying markdown directly. Otoh I wonder which editors this would be.

```

## Try CommonMark

[You💬][1][ can 💬][2][ try💬][1] CommonMark here.  This dingus is powered by
[commonmark.js](https://github.com/commonmark/commonmark.js), the
JavaScript reference implementation.

[1]:  "Comments"

* @ali: My comment to this...
* @bob: I'd rather...


[2]: mention://cem "Comments:
@cem: actually you can't...
"
```

**Comment by mejo-** [comment]
@max-nextcloud and me discussed this briefly:

- Inline comments should be persisted in markdown (not in DB or yjs document) to
  - make them accessible in markdown readers
  - prevent sync issues when the document source got edited
- Inline comments will most likely mostly be used in the web interface.
  - Focus is to not break legibility of the markdown source. Therefore we tend to persist the comments at the end of the document, like footnotes.
- If we support comments for ranges of comments, we probably should not support nested or overlapping comments for the sake of simplicity. We have to ensure proper error handling in case of broken/nested/overlapping comments syntax in the markdown though.
- Syntax idea is to use link references with a link title that start with `💬`
- UI:
  - Use a bubble like the link bubble to display and edit comments
  - Allow to add comments via menu bar and keyboard shortcut
  - Visualize existing comments with highlighting/underlining the commented range. When putting cursor inside the comment, the comment/link bubble will open

**Comment by max-nextcloud** [comment]
I've been pondering this some more. Haven't really made up my mind yet - so throwing some random thoughts out.

## Using links to denote comments

We could use links to anchors in the file to denote comments like mentioned above.
```markdown
[Text I want to comment on](#comment1)
... more text...
... end of the actual content

<div id="comment1">
@max-nextcloud : my comment
@someone-else: their comment
</div>
```

Pros:
* This could enable a meaningful rendering in other markdown tools. 
* Uses standard markdown syntax + some minimal html - which is also valid in markdown.
* Would allow replying to comments as each comment has it's own section.

Cons:
* Comments are separated from what they comment on.
* We might be left with dangling comments after the original content has been removed.
* We might be left with links pointing nowhere if the comment in question has been removed
* Cannot comment on links or text that contains links, comments cannot overlap.
* What markup is valid in comments? Whole can of worms there.

## CriticMarkup

Pros:
* Documented standard.
* Allows comments on links.
* Comments are inside the paragraph they refer to.
* Clearly limits content of comments to inline content.

Cons:
* Is any other tool actually using this? I saw discussions on [hedgedoc](https://github.com/hedgedoc/hedgedoc/issues/2879), [hackmd]( https://github.com/hackmdio/codimd/issues/917) and [cryptpad](https://github.com/cryptpad/cryptpad/issues/742) but none of them implemented it.
* If not parsed syntax can be somewhat confusing.
* Does not work well for replying to comments.


## Using our existing comments on files system

We do have comments in Nextcloud. It's not used so often I think - but we could improve that by embedding it in text.

Pros:
* Easily stores metadata about comments (author, timestamp).
* Permissions already build in - i can edit or delete my own comments.
* Would work on any file type supported by text - not just markdown.
* Keeping the meta conversation separate from the file content.

Cons:
* No replies either.
* Would need extension to support referring to specific text.
* Would get out of sync if file content is updated outside of text.

**Comment by manicmarvin** [comment]
My two pence [disclaimer: not really contributing code to this repo].




> ## Using links to denote comments
> Cons:
>  - Comments are separated from what they comment on.
>  - We might be left with dangling comments after the original content has been removed.
>  - We might be left with links pointing nowhere if the comment in question has been removed
>

Apart from the other cons, IMO these issues pose serious obstacles for large groups where not everyone is familiar with the comment protocol. Those editing files primarily in an external editor are almost sure to break others comments because it will be easy to miss that the comment even exists or how it works. 


> 
> ## CriticMarkup
> Pros:
>  - Comments are inside the paragraph they refer to.
>  -  Clearly limits content of comments to inline content.
>

These address the issue I mentioned above, it's harder to miss new unique syntax and will probably ensure unfamiliar users pause and consider before making changes.

> Cons:
> 
>  - Is any other tool actually using this? I saw discussions on [hedgedoc](https://github.com/hedgedoc/hedgedoc/issues/2879), [hackmd](https://github.com/hackmdio/codimd/issues/917) and [cryptpad](https://github.com/cryptpad/cryptpad/issues/742) but none of them implemented it.
>

even if no one else chose this path, it seems to me to be the best path for NC, given how the files within it are edited by a different people using a plethora of different editors.

>     * If not parsed syntax can be somewhat confusing.
>

A pro, not a con, IMO.


> ## Using our existing comments on files system
> Cons:
> 
>    - Would need extension to support referring to specific text.
>    -  Would get out of sync if file content is updated outside of text.

These two issues make this path non-viable, IMO. I am constantly thinking of the way nextcloud is often used -- in large heterogenous groups or organisations where not everyone is familiar with all features of all tools.

**Comment by marcoambrosini** [comment]
@nextcloud/designers anyone interested in picking this up?

**Comment by juliushaertl** [comment]
From the discussion with @marcoambrosini and @max-nextcloud:

We currently see 3 options
- footnotes to comment on positions
- footnoites comment on paragraph (ending)
- comment on ranges which would not be possible in markdown

- we could aim for more features if we'd introduce a new file format, but not markdown compatible, likely not an option
- likely preferrable to have comments on paragraphs (like on github diffs)
- suggestions could be like on github

## Stretch goals

- would be interesting to merge with existing comments app (maybe as a stretch goal)
  - still preferred to have the comments in the markdown file, otherwise duplicate places to hold data
- replies (stretch goal)
- suggest changes 
- approve reject

**Comment by susnux** [comment]
> * footnotes to comment on positions

I would rather go with HTML comments as a syntax that way we would also solve the problem that text does not preserve HTML comments

**Comment by marcoambrosini** [comment]
@juliushaertl could we schedule a kickoff meeting with stakeholders to scope down this feature?

**Comment by susnux** [comment]
> The benefit of footnotes would be that it does not impact readability of the plain markdown too much, with html comments that is probably getting quite messy if there are a couple of comments on the same line

Thats true, I was just hoping that text would now preserve my comments 😄

**Comment by max-nextcloud** [comment]
I started #6478 to discuss the approach we're currently pushing forward: Comments - but not in line but rather on a per block element basis.

**Comment by juliusknorr** [comment]
@max-nextcloud Let me close this in favor of https://github.com/nextcloud/text/issues/6478 then. Don't see a point in keeping two issues open if we're aiming for one implementation of comments ;)

**Comment by juliusknorr** [comment]
Or wait, since this probably has more detailed description. Let's use this one. I'll copy the current state of discussion over.

**Comment by juliusknorr** [comment]
Copying over design spec draft from @nimishavijay (https://github.com/nextcloud/text/issues/6478#issuecomment-2385154856)

---

## References

<details><summary>Google docs</summary>

![42a0cdf3-4d18-8022-8005-0c92e4168cf5](https://github.com/user-attachments/assets/e3a79d32-d214-48e1-853d-588d23105906)</details>
<details><summary>Notion</summary>

![2d265c62-6e48-80b8-8005-0c92ddf5fe3b](https://github.com/user-attachments/assets/fcc991f0-cb75-4761-a45d-0b383a0ebd07)</details>
<details><summary>Confluence</summary>

![2d265c62-6e48-80b8-8005-0c954d7669cb](https://github.com/user-attachments/assets/1ce3fb23-5ed3-4ff7-bd61-a2184213a7d6)</details>

## Design specs

### Existing comments

![image](https://github.com/user-attachments/assets/9e8f61f3-34a3-4dc9-b203-11fe3282502c)

- [ ] Highlight the text (not the block, that looks too, well... blocky 😀) in `color-primary-element-light` at 0.5 opacity
  - [ ] Nice-to-have: a `border-bottom: 2px solid var(--color-primary-element)` also with 0.5 opacity
- [ ] On hover: highlight `opacity: 1`
- [ ] On click: highlight `opacity: 1` and the comment opens

![image](https://github.com/user-attachments/assets/ba15c0f5-6d5e-412f-aee1-ba05e0dffbcf)

- [ ] Comment box: `max-width: 300px`
- [ ] Show avatar (34px), name, and comment with actions in a 3dot menu
  - [ ] nice-to-have: timestamp
- [ ] Actions in the 3 dot menu: Delete if it's user's comment
  - [ ] nice-to-have: "Mark as resolved" if it is not user's comment

### New comments

![image](https://github.com/user-attachments/assets/36e3d6bd-cea0-4c68-884e-87d5e788c5f7)

- [ ] When the cursor is in a block or when you hover on a block, a comment button should appear (TBD: where and how it looks)
- [ ] When you click on comment button:
  - [ ] text in block gets highlight (opacity 1) 
  - [ ] New comment element opens

![image](https://github.com/user-attachments/assets/f38c5745-ec3e-4a65-b5c0-e9229f98bfaf)

New comment element:
- [ ] Show user's avatar (24px), name, input field, comment and cancel buttons
- [ ] When you click on comment, the comment gets added and is shown as normal

## Mobile view
Mockups TBD
- [ ] Don't show the comment icon next to the block, instead show it on the top bar
- [ ] Existing comments are stickied to right below the block they are associated to
- [ ] Same for new comments, plus the keyboard opens

TBD: 
- [ ] Placement and exact of the comment button next to the block (left side? right side? secondary button? shadow?)
- [ ] Placement of comment action in the top bar 
- [ ] Should there be a "Show/hide comments" toggle in the top bar?
- [ ] Interaction of comments and author colors
- [ ] How to show comments from different people:
  - do we need it for the MVP?
  - different colored highlight? 
  - comment indication with avatar? 
![image](https://github.com/user-attachments/assets/6b19fb74-1af9-4549-bf5e-03b61642eb1f)

What do you think? :)

**Comment by max-nextcloud** [comment]
This is really hard to get right as it adds another layer on top of the document that is completely orthogonal to everything else: 
1. Comments can span multiple blocks (even if we try to have per block comments one can split the block.)
2. Text that is commented upon may contain links.
3. Comments can lead to discussions.  
4. We need to define what is valid content of a comment.

I think we need to conceptually differentiate between the discussion (i.e. a comment and its replies) and the marks on text it discusses (i.e. the highlighted text block). Discussions need to address 3. and 4. while the marks need to account for 1. and 2.

## The discussion

* similar to a heading this would be a separate block on the page that has an anchor - so can be linked to.
* The anchor needs to be prefixed (`d-` most likely) to make sure it does not conflict with css selectors etc.
* For now I would say it's a list that allows for marks and uses mentions to indicate who wrote which line.
* we need to somehow keep track of the id of the discussion in the markdown
* These could be serialized at the end of the file or after the paragraph they refer to. The latter would allow us to just render them on the previous paragraph if there are no marks.

### Possible Syntax:
```
:::discussion d-1
* @max-nextcloud A discussion would be similar to a list - but each item starts with a mention
* @someoneelse And what happens if someone edits the content and messes up the syntax
* @max-nextcloud Then tiptap will try to parse it and remove all content that does not fit the schema.
:::
```

## Marks

* Similar to links they link the comment to the text in question.
* Unlike links though they may be overlapping (in the future).
* We also need to account for overlapping links and comments.

### Possible Syntax:
```
[simple text commented upon](#d-1)

[text with ](#d-1)[link](https://nextcloud.com #d-1)[ commented upon ](#d-1)

[overlapping ](#d-1)[comments](#d-1 #d-2)[ here](#d-2)

```

**Comment by ntnsndr** [comment]
I'm so glad to see this discussion happening. The lack of comments (and suggestions) is the major pain point of the text app for my team. It would make a HUGE difference to have Google Docs-style comment threads.

**Comment by mercury1337** [comment]
Yes, this would be really great to have in my favourite wiki 💘

**Comment by Aveyron-RetD** [comment]
I have seen that there is an obsidian [plugin](https://github.com/Fevol/obsidian-criticmarkup) which is trying to achieve exactly this feature request. They use an improved [criticmarkup syntax](https://github.com/Fevol/criticmarkup-parser/)  but also track changes by user.

The development of the plugin is currently active.

**Comment by codedust** [comment]
The people from [DINUM](https://www.numerique.gouv.fr/) and [ANCT](https://anct.gouv.fr/) are currently also working on this feature in [suitenumerique](https://github.com/suitenumerique/)/[docs](https://github.com/suitenumerique/docs): https://github.com/suitenumerique/docs/issues/156

**Comment by muppeth** [comment]
Wouldn't simply adding Activity tab like when editing the same file outside of collectives app be already a good step forward?
Currently Not being able to discuss the file with others is a big roadblock for a lot of people. Sure inline comments would be much better improvement, but I think Activity would be an easy low hanging fruit.

**Comment by coolidfrez** [comment]
Thanks for pointing me to this existing issue! Took me a while to find this thread, but very happy to discover it!

I was pointing to the same thing as @codedust - the DINUM/ANCT team is also working on this feature in suitenumerique/docs#156. 

I don't really bring anything new to the table, but I'll be happy to communicate about the development of this solution within our institution.

Looking forward to testing it! 👍
